src/mongoc/mongoc-stream-tls.c

```diff
@@ -495,40 +497,131 @@ _mongoc_stream_tls_writev (mongoc_stream_t *stream,
+             /* If we have:
+              *   - buffered bytes already
+              *   - another iovec to send after this one and we don't have more
+              *     bytes to send than the size of the buffer.
+              *
+              * copy into the buffer */
 
-         if (write_ret < 0) {
-            return write_ret;
-         }
+             bytes = BSON_MIN (iov[i].iov_len - iov_pos, buf_end - buf_tail);
 
-         if (expire) {
-            now = bson_get_monotonic_time ();
+             memcpy (buf_tail, iov[i].iov_base + iov_pos, bytes);
+             buf_tail += bytes;
+             iov_pos += bytes;
 
-            if ((expire - now) < 0) {
-               if (write_ret == 0) {
-                  mongoc_counter_streams_timeout_inc();
+             if (buf_tail == buf_end) {
+                /* If we're full, send */
+                buf_iovec.iov_base = buf_head;
+                buf_iovec.iov_len = buf_tail - buf_head;
+
+                child_ret = _mongoc_stream_tls_writev_helper (stream, &buf_iovec, 1,
+                                                              tls->timeout_msec, NULL);
+
+                if (child_ret < 0) {
+                   /* Buffer write failed, just return the error */
+                   return child_ret;
+                }
+
+                ret += child_ret;
+
+                if (child_ret < buf_iovec.iov_len) {
+                   /* we timed out, so send back what we could send */
+
+                   return ret;
+                }
+
+                /* reset all the pointers */
+                buf_tail = buf_head = buf;
+             }
+          } else {
+            /* Didn't buffer, so just write it through */
+
+            write_ret = BIO_write (tls->bio, (char *)iov[i].iov_base + iov_pos,
+                                 (int)(iov[i].iov_len - iov_pos));
+
+            if (write_ret < 0) {
+               return write_ret;
+            }
+
+            if (expire) {
+               now = bson_get_monotonic_time ();
+
+               if ((expire - now) < 0) {
+                  if (write_ret == 0) {
+                     mongoc_counter_streams_timeout_inc();
 #ifdef _WIN32
-                  errno = WSAETIMEDOUT;
+                     errno = WSAETIMEDOUT;
 #else
-                  errno = ETIMEDOUT;
+                     errno = ETIMEDOUT;
 #endif
-                  return -1;
-               }
+                     return ret ? ret : -1;
+                  }
 
-               tls->timeout_msec = 0;
-            } else {
-               tls->timeout_msec = (expire - now) / 1000L;
+                  tls->timeout_msec = 0;
+               } else {
+                  tls->timeout_msec = (expire - now) / 1000L;
+               }
             }
+
+            ret += write_ret;
+            iov_pos += write_ret;
          }
+      }
+   }
+
+   if (buf && buf_head != buf_tail) {
+      /* If we have any bytes buffered, send */
```

Comment:
can you clarify how we would end up with buffered bytes that haven't been written after the loop terminates?